pkg/controller/actions/cacher/cacher.go

```diff
@@ -0,0 +1,68 @@
+package cacher
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/types"
+)
+
+// Renderer is the actual function received from upper layers
+// which generates resources based on the ReconciliationRequest.
+// Resource can be any type, so it returns `any` (or `error`)
+// The upper layers passing the function and can cast it to the expected type.
+type Renderer func(ctx context.Context, rr *types.ReconciliationRequest) (any, error)
+
+type CachingKeyFn func(rr *types.ReconciliationRequest) ([]byte, error)
+
+type Cacher struct {
+	cachingKeyFn    CachingKeyFn
+	cachingKey      []byte
+	cachedResources any
```

Comment:
maybe generics can be used here so we could improve type safety ?

Reply:
> maybe generics can be used here so we could improve type safety ?

I tried :) but opened for suggestions